### PR TITLE
teuthology/suite: pass the base config to scripts

### PR DESCRIPTION
**Pull request by batrick**
This PR is to support: https://github.com/ceph/ceph/pull/60386

Also, this PR fixes a bug: `teuthology/tasks/kernel: default branch should be "distro"`. When the `kernel` dictionary is empty, the `kernel` task should assume `distro`.

### CODE REVIEW

**Comment by batrick** [comment]
> Also, this PR fixes a bug: `teuthology/tasks/kernel: default branch should be "distro"`. When the `kernel` dictionary is empty, the `kernel` task should assume `distro`.

Example: https://pulpito.ceph.com/pdonnell-2024-10-21_14:21:05-fs-wip-pdonnell-testing-20241019.005706-debug-wip-pdonnell-i66704-default-smithi/

File: teuthology/suite/merge.py [hunk 1]
```diff
@@ -115,20 +115,22 @@ def config_merge(configs, suite_name=None, **kwargs):
...
     """
     seed = kwargs.setdefault('seed', 1)
+    base_config = kwargs.setdefault('base_config', {})
     if not isinstance(seed, int):
         log.debug("no valid seed input: using 1")
         seed = 1
     log.debug("configuring Lua randomseed to %d", seed)
     L.execute(f'local math = require"math"; math.randomseed({seed});')
     new_script = L.eval('new_script')
+    base_config_yaml = yaml.safe_load(str(base_config))
```
Comment (zmc) [inline teuthology/suite/merge.py:125]: It's not immediately obvious why we'd need to create a copy of `base_config` here and then again on 133 - could you explain the reasoning?

**Review by zmc — COMMENTED** [review]
Nice! I just have one question

File: teuthology/suite/merge.py (hunk 1 above)
Comment (batrick) [reply teuthology/suite/merge.py:125]: It's to basically enforce a read-only view of the base_config.

Comment (batrick) [reply teuthology/suite/merge.py:125]: I tried to do `copy.deepcopy(base_config)` below but ran into inscrutable recursion errors that didn't seem worth debugging. So I'm just reloading it as a legit yaml dictionary here.

Comment (zmc) [reply teuthology/suite/merge.py:125]: I think I would suggest just doing a `yaml_complete_obj = copy.deepcopy(base_config.to_dict())` on 133 and dropping this statement.

Comment (batrick) [reply teuthology/suite/merge.py:125]: ack

Comment (batrick) [reply teuthology/suite/merge.py]: done

File: teuthology/suite/merge.py [hunk 2]
```diff
@@ -115,6 +115,7 @@ def config_merge(configs, suite_name=None, **kwargs):
     the entire job (config) from the list.
     """
     seed = kwargs.setdefault('seed', 1)
+    base_config = kwargs.setdefault('base_config', {})
```
Comment (zmc) [inline teuthology/suite/merge.py:118]:
```suggestion
    base_config = kwargs.setdefault('base_config', JobConfig())
```
And add `from teuthology.config import JobConfig` up top

Comment (batrick) [reply teuthology/suite/merge.py]: thanks, done

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2024-10-23_02:34:42-fs-wip-pdonnell-testing-20241019.005706-debug-wip-pdonnell-i66704-default-smithi

**Comment by batrick** [comment]
@zmc good to go?

**Review by zmc — APPROVED** [review]

**Pull request merged**